### PR TITLE
Temporary fix for Issue #1841

### LINKED ISSUE #1841: Preventing printing overflow for long circuits in terminal with `qml.draw_circuit()`
### Feature details

Calling `qml.drawcircuit` and rendering the output to the terminal results in illegible circuits if the circuit width (*i.e.* the number of text columns used) is greater than the column width of the terminal window. This issue is due to the overflow of text used to visualize the circuit onto the next line. For example:

**Current Behavior:**
```
|------------------terminal width ------------------|
 0: ─H─RZ(1.0)──╭C──────────╭X─RX(3.14)─╭C──────────
 ╭X─RX(3.14)──────┤ ⟨Z⟩
 1: ─H─RZ(0.1)──╰X─╭C───────│──RX(3.14)─╰X─╭C───────
 │──RX(3.14)──────┤
 2: ─H─RZ(0.2)─────╰X─╭C────│──RX(3.14)────╰X─╭C────
 │──RX(3.14)──────┤
 3: ─H─RZ(0.3)────────╰X─╭C─│──RX(3.14)───────╰X─╭C─
 │──RX(3.14)──────┤
 4: ─H─RZ(0.4)───────────╰X─╰C─RX(3.14)──────────╰X─
 ╰C─RX(3.14)──────┤
```
A possible solution would be to detect the width of the terminal prior to printing the circuit. With the width detected, the printing of lines could occur with the circuit length modulo the terminal width to facilitate line skipping as needed. This skipping would be repeated to allow the full circuit to be cleanly rendered in the terminal without the overflow issues. For example:

**Alternative Behavior:**
```
|------------------terminal width ------------------|
 0: ─H─RZ(1.0)──╭C──────────╭X─RX(3.14)─╭C────────── 
 1: ─H─RZ(0.1)──╰X─╭C───────│──RX(3.14)─╰X─╭C───────
 2: ─H─RZ(0.2)─────╰X─╭C────│──RX(3.14)────╰X─╭C────
 3: ─H─RZ(0.3)────────╰X─╭C─│──RX(3.14)───────╰X─╭C─
 4: ─H─RZ(0.4)───────────╰X─╰C─RX(3.14)──────────╰X─

 0: ─╭X─RX(3.14)──────┤⟨Z⟩
 1: ─│──RX(3.14)──────┤
 2: ─│──RX(3.14)──────┤
 3: ─│──RX(3.14)──────┤
 4: ─╰C─RX(3.14)──────┤
```

Provided there's interest, I'd be happy to send a pull request with this feature (although I may need assistance testing it on Linux and Windows operating systems). 

### Implementation

To detect the terminal size:

```python
import shutil
cols, _ = shutil.get_terminal_size()
```

Note that `shutil` is part of the standard library and would not require an additional dependency. 

It seems to me this could be integrated fairly trivially into `qml.circuit_drawer.draw()` (below), but perhaps this could be confirmed by a developer more familiar with the code base. 

```python
    def draw(self):
        """Draw the circuit diagram.
        Returns:
            str: The circuit diagram
        """
        rendered_string = ""

        # extract the wire labels as strings and get their maximum length
        wire_names = []
        padding = 0
        for i in range(self.full_representation_grid.num_wires):
            wire_name = str(self.active_wires.labels[i])
            padding = max(padding, len(wire_name))
            wire_names.append(wire_name)

        for i in range(self.full_representation_grid.num_wires):
            # format wire name nicely
            wire = self.full_representation_grid.wire(i)
            s = " {:>" + str(padding) + "}: {}"

            rendered_string += s.format(wire_names[i], 2 * self.charset.WIRE)

            for s in wire:
                rendered_string += s

            rendered_string += "\n"

        for symbol, cache in [
            ("U", self.representation_resolver.unitary_matrix_cache),
            ("H", self.representation_resolver.hermitian_matrix_cache),
            ("M", self.representation_resolver.matrix_cache),
        ]:
            for idx, matrix in enumerate(cache):
                rendered_string += "{}{} =\n{}\n".format(symbol, idx, matrix)

        return rendered_string
```

### How important would you say this feature is?

1: Not important. Would be nice to have.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by greenstick**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

In reference to Issue #1841. From the discussion, this is a temporary fix to the text overflow issue when printing a circuit to the terminal. 

**Description of the Change:**

The code works by allowing the user to set a `max_length` keyword argument when calling `CircuitDrawer.draw()`. Updates have been made to the `draw` method of the `CircuitDrawer` class and functions referencing it to allow for the `max_length` keyword arg. A unit test for the parameter has also been created.

**Benefits:**
Legible circuits in the CLI as we aspire to polynomial length circuits! Step aside, AC0.  

**Possible Drawbacks:**
- This code isn't a long term solution. Its a quick usability fix. From the discussion in Issue #1841, the code may soon be replaced by a refactored `CircuitDrawer` class. 
- I've unfortunately been unable to run the test suite to verify this code and the unit test due to a number of other errors when running the suite. I have developed this with attention to detail around not breaking existing tests. 

**Related GitHub Issues:**
Closes #1841.

### CODE REVIEW

**Comment by josh146** [comment]
Thanks @greenstick! I've approved the CI workflow, so you should be able to use this PR to check that all the tests pass 🙂 Let us know if you have any questions, or when this is ready for review.

**Comment by greenstick** [comment]
CodeFactor is failing on the too many arguments flag (6/5) following the update to `qml.transforms.draw()`. Any recommendations on how to manage this without a larger refactor would be great.

**Comment by josh146** [comment]
this one is fine to disable, as we treat it more like a warning 🙂 

At the top of the file, below the module docstring, simply add the comment

```python
# pylint: disable=too-many-arguments
```

File: tests/circuit_drawer/test_circuit_drawer.py [hunk 1]
```diff
@@ -644,6 +644,35 @@ def drawn_qubit_circuit_with_interesting_wires():
     )
 
 
+@pytest.fixture
```
Comment (albi3ro) [inline tests/circuit_drawer/test_circuit_drawer.py:647]: I don't think there's a need for fixtures in this case. Our test suite can tend to overuse them, but that's a work in progress problem.  They are only used in one test, and they aren't something that requires post-test cleanup.

Also, we've found including the information as close together as possible aides debugging down the line when something goes wrong.  If someone can just copy and paste the code within one test function, without searching all over to find relevant variables, debugging becomes much easier.

File: tests/circuit_drawer/test_circuit_drawer.py [hunk 2]
```diff
@@ -644,6 +644,35 @@ def drawn_qubit_circuit_with_interesting_wires():
...
+
+
+@pytest.fixture
+def drawn_qubit_circuit_with_max_length_kwdarg():
+    """The renderedd circuit representation of the above qubit circuit with max_length set to 30"""
+    return (
+        " 0: ──H──RX(0)────RY(0)────RZ\n"
+        + " 1: ──H──RX(0.1)──RY(0.1)──RZ\n"
+        + " 2: ──H──RX(0.2)──RY(0.2)──RZ\n"
+        + "\n"
```
Comment (albi3ro) [inline tests/circuit_drawer/test_circuit_drawer.py:669]:
```suggestion
        + " \n"
```

This should get the test passing. Or you could modify the code so this space does appear.  Don't think it matters too much.

**Review by albi3ro — COMMENTED** [review]
You'll also want to test the integrated `qml.draw` transform.  See it's test file at [tests/transforms/test_draw.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/transforms/test_draw.py)

**Comment by greenstick** [comment]
Manages to get the test suite working so, with a little luck, the last commit should pass all the tests.

File: pennylane/circuit_drawer/circuit_drawer.py [hunk 3]
```diff
@@ -81,6 +83,9 @@ def __init__(
             # as active wires
             self.active_wires = wires.all_wires([wires, self.active_wires])
 
+        # We add a -2 character offset to account for some downstream formatting
+        self.max_length = max_length - 2 if type(max_length) == int else None
```
Comment (albi3ro) [inline pennylane/circuit_drawer/circuit_drawer.py:87]:
```suggestion
        self.max_length = max_length - 2 if max_length is not None else None
```

**Review by albi3ro — APPROVED** [review]
Thanks @greenstick ! 🎉 🚀

**Comment by greenstick** [comment]
Noticed that the code block to enable this feature in the `CircuitDrawer` was out of order to render the matrix after the circuit.

**Comment by josh146** [comment]
@greenstick feel free to open another PR, unfortunately this one has been merged and cannot be re-opened :)

**Comment by CatalinaAlbornoz** [comment]
Hi @greenstick, thank you for contributing to PennyLane! Since your contribution was merged to PennyLane during November you have earned a Quantum Carnival ticket! Please send me a private message on [Slack](https://u.strawberryfields.ai/slack/) or [Discord](https://discord.com/invite/JqVGmpkP96) so that I can send you your ticket. Congratulations!

**Pull request merged**